src/terms.rs

```diff
@@ -0,0 +1,529 @@
+        // doing canonicalization
+        // (and through the view merge function, rebuilding)
+
+        // Actually, we do one rule per column for
+        // efficiency- joining on one column at a time
+        // is more efficient
+        for i in 0..(fdecl.schema.input.len() + 1) {
+            let var_type = if i == fdecl.schema.input.len() {
+                types.output.clone()
+            } else {
+                types.input[i].clone()
+            };
+            let current = if i == fdecl.schema.input.len() {
+                "lhs".to_string()
+            } else {
+                child(i)
+            };
+            let current_updated = if i == fdecl.schema.input.len() {
+                lhs_updated.clone()
+            } else {
+                child_parent(self, i)
+            };
+            let new_lhs = self.fresh_var();
+            let new_term_for_view = if types.output.is_eq_sort() {
+                format!(
+                    "(let {new_lhs} ({op_name} {children_updated}))
+                         {}",
+                    self.union(
+                        types.output.name(),
+                        &new_lhs.to_string(),
+                        lhs_updated.as_str(),
+                    )
+                )
+            } else {
+                "".to_string()
+            };
+            if var_type.is_eq_sort() {
+                let rule = format!(
+                    "(rule ((= lhs ({view_name} {children}))
+                            (!= {current} {current_updated}))
+                     (
+                      {new_term_for_view}
+                      (delete ({view_name} {children}))
+                      (set ({view_name} {children_updated}) {lhs_updated})
+                     )
+                      :ruleset {})",
+                    self.rebuilding_ruleset_name()
+                );
+                res.extend(self.desugar().parse_program(&rule).unwrap());
+            }
+        }
+
+        res
+    }
+
+    /// Instrument fact replaces terms with looking up
+    /// canonical versions in the view.
+    /// It also needs to look up references to globals.
+    fn instrument_fact(&mut self, fact: &NormFact) -> Fact {
+        match fact {
+            NormFact::Assign(lhs, NormExpr::Call(head, body)) => {
+                let func_type = self
+                    .type_info()
+                    .lookup_expr(self.current_ctx, &NormExpr::Call(*head, body.clone()))
+                    .unwrap();
+                if func_type.is_datatype {
+                    let view_name = self.view_name(*head);
+                    NormFact::Assign(*lhs, NormExpr::Call(view_name, body.clone()))
+                } else {
+                    fact.clone()
+                }
+            }
+            _ => fact.clone(),
+        }
+        .map_use(&mut |var| {
+            let vtype = self.type_info().lookup(self.current_ctx, var).unwrap();
+            if !self.type_info().is_global(var) {
```

Comment:
hmm I'm confused. I thought globals should always be canonical?

Reply:
See previous comment